### ISSUE TITLE
Listeners not working?

### ISSUE THREAD

**Issue by BenRomberg** [jhalterman/failsafe#36]
First of all, great framework! Lightweight, super simple to use, yet feature-rich. Enjoying it so far!

Sadly I can't get Listeners to work. Trying the `onRetry` and `onFailedAttempt` Listeners with more than one argument, with the `retryIf` condition failing. Some code:

```
    private static final RetryPolicy RETRY_POLICY = new RetryPolicy()
            .retryIf((CachedResponse response) -> response == null || response.getStatusFamily() != Family.SUCCESSFUL)
            .retryOn(Exception.class).withDelay(3, TimeUnit.SECONDS).withMaxRetries(3);
// ...
CachedResponse response = Failsafe.<Response> with(RETRY_POLICY)
                .onFailedAttempt((failedResponse, exception, context) -> {
                    System.err.println("response: " + failedResponse);
                    System.err.println("exception: " + exception);
                    }).get(() -> new CachedResponse(target.request().get()));
```

Some scenarios that I expect to work vs. the ones that actually work:
- [ ] Response fails (i.e. `retryIf` condition fails) with 3 arguments `onFailedAttempt` / `onRetry`
- [ ] Response fails (i.e. `retryIf` condition fails) with 2 arguments `onFailedAttempt` / `onRetry`
- [x] Response fails (i.e. `retryIf` condition fails) with 1 argument `onFailedAttempt` / `onRetry` (but argument `failure` is always `null`, which is correct but not helpful)
- [x] Failure (i.e. `retryOn` condition fails) with 3 arguments `onFailedAttempt` / `onRetry`
- [x] Failure (i.e. `retryOn` condition fails) with 2 arguments `onFailedAttempt` / `onRetry`
- [x] Failure (i.e. `retryOn` condition fails) with 1 argument `onFailedAttempt` / `onRetry`

Let me know if you can reproduce it or if I'm doing something wrong on my side. Thanks!

**Comment by jhalterman** [comment]
@BenRomberg Thanks for filing this. This was indeed a bug where the failed attempt listener would not always be called the final time. For example, with 3 retries as in your example, it might only be called 3 times, not 4 as expected.

`onRetry` is a little different though since it only gets called **before** a retry happens. So with 3 retries it would only be called 3 times, whereas `onFailedAttempt` should be called 4 times (since the initial execution failed, plus the 3 retry attempts).

Let me know if that doesn't describe the problem you were having. Also, let me know if you can try the current code in master and if that resolves your issue.

**Comment by BenRomberg** [comment]
The problem I had was that the listener was not called at all in the 2 scenarios I described above, even though there were 3 failing retries.

I'll try out `master` once I get to it, probably by the end of next week (it's for a side project I only work on Fridays). If it still doesn't work, I'll provide a failing test that highlights the issue. Thanks for the prompt response!

**Comment by jhalterman** [comment]
@BenRomberg I just updated a test case for this issue, and I think it matches the failing scenario you describe:

https://github.com/jhalterman/failsafe/blob/4bf02f7fcfc5737b4ffe5f03ca2cd4ab591d430e/src/test/java/net/jodah/failsafe/issues/Issue36.java

Right now the test passes. If you do get a chance to come up with a failing test for what you're seeing, that would be great.

**Comment by BenRomberg** [comment]
@jhalterman: Finally had the time to try it out. Either it still doesn't work or I'm doing something wrong.

Below are 4 tests that I would expect to pass:

```
    @Test
    public void failedAttemptListener_WithFailedResponses_ShouldBeCalled() throws Exception {
        @SuppressWarnings("unchecked")
        RetryPolicy policy = new RetryPolicy().retryIf((Boolean response) -> response != null && response)
                .retryOn(Exception.class).withDelay(1, TimeUnit.SECONDS).withMaxRetries(3);
        AtomicInteger listenerCallbacks = new AtomicInteger();
        Failsafe.<Response> with(policy)
                .onFailedAttempt((failedResponse, exception, context) -> listenerCallbacks.incrementAndGet())
                .get(() -> false);
        assertThat(listenerCallbacks.get()).isEqualTo(4);
    }

    @Test
    public void retryListener_WithFailedResponses_ShouldBeCalled() throws Exception {
        @SuppressWarnings("unchecked")
        RetryPolicy policy = new RetryPolicy().retryIf((Boolean response) -> response != null && response)
                .retryOn(Exception.class).withDelay(1, TimeUnit.SECONDS).withMaxRetries(3);
        AtomicInteger listenerCallbacks = new AtomicInteger();
        Failsafe.<Response> with(policy)
                .onRetry((failedResponse, exception, context) -> listenerCallbacks.incrementAndGet()).get(() -> false);
        assertThat(listenerCallbacks.get()).isEqualTo(3);
    }

    @Test
    public void failedAttemptListener_WithExceptions_ShouldBeCalled() throws Exception {
        @SuppressWarnings("unchecked")
        RetryPolicy policy = new RetryPolicy().retryIf((Boolean response) -> response != null && response)
                .retryOn(Exception.class).withDelay(1, TimeUnit.SECONDS).withMaxRetries(3);
        AtomicInteger listenerCallbacks = new AtomicInteger();
        try {
            Failsafe.<Response> with(policy)
                    .onFailedAttempt((failedResponse, exception, context) -> listenerCallbacks.incrementAndGet())
                    .get(() -> {
                        throw new RuntimeException();
                    });
        } catch (RuntimeException e) {
        }
        assertThat(listenerCallbacks.get()).isEqualTo(4);
    }

    @Test
    public void retryListener_WithExceptions_ShouldBeCalled() throws Exception {
        @SuppressWarnings("unchecked")
        RetryPolicy policy = new RetryPolicy().retryIf((Boolean response) -> response != null && response)
                .retryOn(Exception.class).withDelay(1, TimeUnit.SECONDS).withMaxRetries(3);
        AtomicInteger listenerCallbacks = new AtomicInteger();
        try {
            Failsafe.<Response> with(policy)
                    .onRetry((failedResponse, exception, context) -> listenerCallbacks.incrementAndGet()).get(() -> {
                        throw new RuntimeException();
                    });
        } catch (RuntimeException e) {
        }
        assertThat(listenerCallbacks.get()).isEqualTo(3);
    }
```

Only the last 2 pass (using either 0.8.3, 0.9.1 or 0.9.2-SNAPSHOT from `master`) when the retry is triggered by an exception, but not if it's triggered by an unsatisfying result from the callback (from the `retryIf` condition in the policy). As you've used only the retry-on-exception case in your test, it passes, but what I'm talking about is the `retryIf` condition not triggering the listener correctly.

**Comment by jhalterman** [comment]
@BenRomberg Thanks for posting this. I think it might just be a problem with the `retryIf` statement. Should it be `.retryIf((Boolean response) -> response != null && !response)` ? We want to retry when response is `false`.

**Comment by BenRomberg** [comment]
@jhalterman: You're right, I mixed that up. I want it to retry 3 times in the test. But even after changing it, the tests still fail.

**Comment by jhalterman** [comment]
@BenRomberg Do you have any more details on what's going on? I just added your test cases (without the delay), and they're passing:

ef79e251d26f8830df99e6cd0512fcc0ee4ebee8

**Comment by BenRomberg** [comment]
@jhalterman: I finally figured it out. It was because of the wrong generic type `<Response>` I was calling the `Failsafe.<Response> with` method with. I changed the types during development and forgot to adjust that one. I think it was necessary in one of the earlier versions of Failsafe, nowadays it's not needed at all and I can just call `Failsafe.with(...)` without having to bind an explicit type.

Thanks for your support!
